google/cloud/spanner_dbapi/parse_utils.py

```diff
@@ -402,107 +502,13 @@ def ensure_where_clause(sql):
         for token in sqlparse.parse(sql)[0]
     ):
         return sql
-    return sql + " WHERE 1=1"
 
+    if os.environ.get("RUNNING_SPANNER_BACKEND_TESTS") == "1":
+        return sql + " WHERE 1=1"
 
-SPANNER_RESERVED_KEYWORDS = {
-    "ALL",
-    "AND",
-    "ANY",
-    "ARRAY",
-    "AS",
-    "ASC",
-    "ASSERT_ROWS_MODIFIED",
-    "AT",
-    "BETWEEN",
-    "BY",
-    "CASE",
-    "CAST",
-    "COLLATE",
-    "CONTAINS",
-    "CREATE",
-    "CROSS",
-    "CUBE",
-    "CURRENT",
-    "DEFAULT",
-    "DEFINE",
-    "DESC",
-    "DISTINCT",
-    "DROP",
-    "ELSE",
-    "END",
-    "ENUM",
-    "ESCAPE",
-    "EXCEPT",
-    "EXCLUDE",
-    "EXISTS",
-    "EXTRACT",
-    "FALSE",
-    "FETCH",
-    "FOLLOWING",
-    "FOR",
-    "FROM",
-    "FULL",
-    "GROUP",
-    "GROUPING",
-    "GROUPS",
-    "HASH",
-    "HAVING",
-    "IF",
-    "IGNORE",
-    "IN",
-    "INNER",
-    "INTERSECT",
-    "INTERVAL",
-    "INTO",
-    "IS",
-    "JOIN",
-    "LATERAL",
-    "LEFT",
-    "LIKE",
-    "LIMIT",
-    "LOOKUP",
-    "MERGE",
-    "NATURAL",
-    "NEW",
-    "NO",
-    "NOT",
-    "NULL",
-    "NULLS",
-    "OF",
-    "ON",
-    "OR",
-    "ORDER",
-    "OUTER",
-    "OVER",
-    "PARTITION",
-    "PRECEDING",
-    "PROTO",
-    "RANGE",
-    "RECURSIVE",
-    "RESPECT",
-    "RIGHT",
-    "ROLLUP",
-    "ROWS",
-    "SELECT",
-    "SET",
-    "SOME",
-    "STRUCT",
-    "TABLESAMPLE",
-    "THEN",
-    "TO",
-    "TREAT",
-    "TRUE",
-    "UNBOUNDED",
-    "UNION",
-    "UNNEST",
-    "USING",
-    "WHEN",
-    "WHERE",
-    "WINDOW",
-    "WITH",
-    "WITHIN",
-}
+    raise ProgrammingError(
+        "Cloud Spanner requires a WHERE clause when executing DELETE or UPDATE query"
+    )
```

Comment:
Apologies if I misunderstood something here.

I believe this clause is added by default because Django will construct  DELETE and UPDATE queries that don't have the WHERE clause as this is not a requirement in other databases. It ensure that customers can smoothly transition from their existing applications without having to write raw queries with the WHERE clause.

The clause wasn't added just to get the tests to pass.

Reply:
@skuruppu, hiding this useful restriction doesn't look right. But I see the point in smooth transition as well. Do you mean that only for Django? Or we should consider that any user can have queries without WHERE, so this dummy should be kept for the whole connection API? (we have `spanner_dbapi` directory, which seems to be intended to be default connection API - it's from where I'm proposing to erase the dummy, and `django_spanner` directory, where Django-specific parts are held).

The main thing that makes me nervous is that queries can be big and branched (for example, UNIONS are little bit hard to interpret, as WHERE clause can be relative to the last of the union parts, but also can be relative to the result of the UNION on the whole, and they both will be at the about same place). I'm not sure if there is a way to check 100% of variants with several regexes, and insert this dummy correctly. So, I think it'll be exploding things from time to time.

Reply:
I agree that it may be good to notify the user about this restriction from the connection API. Is it possible to instead modify the Django implementation to add the WHERE clause in the generated query? I think this is where it matters the most in order to support users migrating from other databases.